librz/include/rz_type.h

```diff
@@ -0,0 +1,238 @@
+	char *name;
+	char *type; // Used by typedef, atomic type, enum
+	ut64 size; // size of the whole type in bits
+	RzBaseTypeKind kind;
+	union {
+		RzBaseTypeStruct struct_data;
+		RzBaseTypeEnum enum_data;
+		RzBaseTypeUnion union_data;
+	};
+} RzBaseType;
+
+typedef struct rz_type_enum {
+	const char *name;
+	const char *val;
+} RzTypeEnum;
+
+#ifdef RZ_API
+
+RZ_API RzTypeDB *rz_type_db_new();
+RZ_API void rz_type_db_free(RzTypeDB *typedb);
+RZ_API void rz_type_db_load_sdb(RzTypeDB *typedb, const char *dpath);
+RZ_API void rz_type_db_purge(RzTypeDB *typedb);
+RZ_API void rz_type_db_set_bits(RzTypeDB *typedb, int bits);
+RZ_API void rz_type_db_set_os(RzTypeDB *typedb, const char *os);
+RZ_API void rz_type_db_set_cpu(RzTypeDB *typedb, const char *cpu);
+RZ_API void rz_type_db_set_endian(RzTypeDB *typedb, bool big_endian);
+RZ_API char *rz_type_db_kuery(RzTypeDB *typedb, const char *query);
+
+RZ_API const char *rz_type_db_get(RzTypeDB *typedb, const char *name);
+RZ_API bool rz_type_db_set(RzTypeDB *typedb, ut64 at, const char *field, ut64 val);
+RZ_API bool rz_type_db_del(RzTypeDB *typedb, RZ_NONNULL const char *name);
+
+RZ_API void rz_type_db_init(RzTypeDB *typedb, const char *dir_prefix, const char *arch, int bits, const char *os);
+
+// Base types
+
+RZ_API void rz_type_base_type(const RzTypeDB *typedb, const RzBaseType *type);
+RZ_API void rz_type_base_type_free(RzBaseType *type);
+RZ_API RzBaseType *rz_type_base_type_new(RzBaseTypeKind kind);
+
+RZ_API void rz_type_base_enum_case_free(void *e, void *user);
+RZ_API void rz_type_base_struct_member_free(void *e, void *user);
+RZ_API void rz_type_base_union_member_free(void *e, void *user);
+
+RZ_API RzBaseType *rz_type_db_get_base_type(RzTypeDB *typedb, const char *name);
+RZ_API void rz_type_db_save_base_type(const RzTypeDB *typedb, const RzBaseType *type);
+
+/* ctype */
+// Parses strings like "const char * [0x42] const * [23]" to RzTypeCTypeType
+
+typedef struct rz_type_ctype_t RzTypeCType;
+
+typedef enum {
+	RZ_TYPE_CTYPE_TYPE_KIND_IDENTIFIER,
+	RZ_TYPE_CTYPE_TYPE_KIND_POINTER,
+	RZ_TYPE_CTYPE_TYPE_KIND_ARRAY
+} RzTypeCTypeTypeKind;
+
+typedef enum {
+	RZ_TYPE_CTYPE_IDENTIFIER_KIND_UNSPECIFIED,
+	RZ_TYPE_CTYPE_IDENTIFIER_KIND_STRUCT,
+	RZ_TYPE_CTYPE_IDENTIFIER_KIND_UNION,
+	RZ_TYPE_CTYPE_IDENTIFIER_KIND_ENUM
+} RzTypeCTypeTypeIdentifierKind;
+
+typedef struct rz_type_ctype_type_t RzTypeCTypeType;
+struct rz_type_ctype_type_t {
+	RzTypeCTypeTypeKind kind;
+	union {
+		struct {
+			RzTypeCTypeTypeIdentifierKind kind;
+			char *name;
+			bool is_const;
+		} identifier;
+		struct {
+			RzTypeCTypeType *type;
+			bool is_const;
+		} pointer;
+		struct {
+			RzTypeCTypeType *type;
+			ut64 count;
+		} array;
+	};
+};
+
+RZ_API RzTypeCType *rz_type_ctype_new(void);
+RZ_API void rz_type_ctype_free(RzTypeCType *ctype);
+RZ_API RzTypeCTypeType *rz_type_ctype_parse(RzTypeCType *ctype, const char *str, char **error);
+RZ_API void rz_type_ctype_type_free(RzTypeCTypeType *type);
+
+/* c */
```

Comment:
are there unit tests for all these functions?

Reply:
No, I plan to add them during the Tree-Sitter refactoring. Please remember this is only a first step in "Types" project: https://github.com/rizinorg/rizin/issues/275